### PR TITLE
[10.0][MIG] partner_second_lastname: Migration to 10.0

### PR DESCRIPTION
**Pull request by pedrobaeza**
Migration commit is dfdc556a5bb5bb297b69deb23faecd1201ca3283

### CODE REVIEW

**Comment by pedrobaeza** [comment]
cc @Tecnativa

File: partner_second_lastname/README.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+Credits
+=======
+
+Contributors
+------------
+
+* `Grupo ESOC <http://grupoesoc.es>`_:
+    * `Jairo Llopis <mailto:j.llopis@grupoesoc.es>`_.
+* `Antiun Ingeniería S.L. <http://www.antiun.com>`_:
+    * `Antonio Espinosa <mailto:antonioea@antiun.com>`_.
```
Comment (yajo) [inline partner_second_lastname/README.rst]: All these credits should be updated.

File: partner_second_lastname/models/res_partner.py [hunk 2]
```diff
@@ -96,17 +97,18 @@ def _get_inverse_name(self, name, is_company=False):
...
             super(ResPartner, self)._check_name()
         except exceptions.EmptyNamesError as error:
-            if not self.lastname2:
-                raise error
+            for partner in self:
+                if not partner.lastname2:
+                    raise error
 
-    @api.one
+    @api.multi
```
Comment (yajo) [inline partner_second_lastname/models/res_partner.py]: It's good to remove the `@api.one` calls, just remember than in v10 `@api.multi` is the same as removing the decorator.

**Review by yajo — APPROVED** [review]

**Review by yajo — APPROVED** [review]
Over-approved now!

File: partner_second_lastname/README.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+=======================
+Partner second lastname
+=======================
+
+This module was written to extend the functionality of ``partner_firstname`` to
+support having a second lastname for contact partners.
+
+In some countries, it's important to have a second last name for contacts.
+
+Contact partners will need to fulfill at least one of the name fields
```
Comment (lasley) [inline partner_second_lastname/README.rst]: s/fulfill/fill

File: partner_second_lastname/README.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+separately (maybe from other form), this module will try to guess the best
+match for your input and split it between firstname, lastname and second
+lastname using an inverse function.
+
+If you can, always enter it manually please. Automatic guessing could fail for
+you easily in some corner cases.
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/134/8.0
```
Comment (lasley) [inline partner_second_lastname/README.rst]: 10.0

File: partner_second_lastname/README.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues
+<https://github.com/OCA/partner-contact/issues>`_. In case of trouble, please
+check there if your issue has already been reported. If you spotted it first,
+help us smashing it by providing a detailed and welcomed feedback `here
+<https://github.com/OCA/partner-contact/issues/new?body=module:%20partner_second_lastname%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (lasley) [inline partner_second_lastname/README.rst]: old template

File: partner_second_lastname/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    "name": "Partner second last name",
+    "summary": "Have split first and second lastnames",
+    "version": "10.0.1.0.0",
+    "license": "AGPL-3",
+    "website": "https://www.tecnativa.com",
+    "author": "Tecnativa, "
+              "Odoo Community Association (OCA)",
+    "category": "Partner Management",
+    "depends": [
+        "partner_firstname"
```
Comment (lasley) [inline partner_second_lastname/__manifest__.py]: trailing comma

File: partner_second_lastname/models/base_config_settings.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+# © 2015 Antiun Ingenieria S.L. - Antonio Espinosa
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import models, api
+
+
+class BaseConfigSettings(models.TransientModel):
+    _inherit = 'base.config.settings'
+
+    def _partner_names_order_selection(self):
```
Comment (lasley) [inline partner_second_lastname/models/base_config_settings.py:11]: `api.model`?

File: partner_second_lastname/models/base_config_settings.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+            BaseConfigSettings, self)._partner_names_order_selection()
+        new_labels = {
+            'last_first': 'Lastname SecondLastname Firstname',
+            'last_first_comma': 'Lastname SecondLastname, Firstname',
+            'first_last': 'Firstname Lastname SecondLastname',
+        }
+        return [(k, new_labels[k]) if k in new_labels else (k, v)
+                for k, v in options]
+
+    @api.multi
```
Comment (lasley) [inline partner_second_lastname/models/base_config_settings.py:22]: `@api.model`? Or `model_cr_context` maybe? Definitely not `multi`

File: partner_second_lastname/models/res_partner.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    @api.depends("firstname", "lastname", "lastname2")
+    def _compute_name(self):
+        """Write :attr:`~.name` according to splitted data."""
+        for partner in self:
+            partner.name = self._get_computed_name(
+                partner.lastname, partner.firstname, partner.lastname2,
+            )
+
+    @api.one
```
Comment (lasley) [inline partner_second_lastname/models/res_partner.py:52]: Deprecated

File: partner_second_lastname/models/res_partner.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    def _inverse_name(self):
+        """Try to revert the effect of :meth:`._compute_name`."""
+        parts = self._get_inverse_name(self.name, self.is_company)
+        # Avoid to hit :meth:`~._check_name` with all 3 fields being ``False``
+        before, after = dict(), dict()
```
Comment (lasley) [inline partner_second_lastname/models/res_partner.py]: `efore, after = {}, {}`

File: partner_second_lastname/models/res_partner.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        for key, value in parts.iteritems():
+            (before if value else after)[key] = value
+        if any([before[k] != self[k] for k in before.keys()]):
+            self.update(before)
```
Comment (lasley) [inline partner_second_lastname/models/res_partner.py]: `.update`?

File: partner_second_lastname/models/res_partner.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        if any([after[k] != self[k] for k in after.keys()]):
+            self.update(after)
```
Comment (lasley) [inline partner_second_lastname/models/res_partner.py:63]: `.update`?

File: partner_second_lastname/models/res_partner.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+
+    @api.constrains("firstname", "lastname", "lastname2")
+    def _check_name(self):
+        """Ensure at least one name is set."""
+        try:
+            super(ResPartner, self)._check_name()
+        except exceptions.EmptyNamesError as error:
+            for partner in self:
+                if not partner.lastname2:
+                    raise error
```
Comment (lasley) [inline partner_second_lastname/models/res_partner.py]: `error` is superfluous - just `raise` inside of the `except` will do the same thing IIRC

**Review by lasley — CHANGES_REQUESTED** [review]

File: partner_second_lastname/models/base_config_settings.py (hunk 7 above)
Comment (pedrobaeza) [reply partner_second_lastname/models/base_config_settings.py:11]: As the overwritten method is by omission `@api.multi`, I let it the same

File: partner_second_lastname/models/base_config_settings.py (hunk 8 above)
Comment (pedrobaeza) [reply partner_second_lastname/models/base_config_settings.py:22]: The same for this: mimicking overwritten method.

File: partner_second_lastname/models/res_partner.py (hunk 9 above)
Comment (pedrobaeza) [reply partner_second_lastname/models/res_partner.py:52]: I know, but I don't want to change this logic right now (not too much time for this) :wink:

File: partner_second_lastname/models/res_partner.py (hunk 11 above)
Comment (pedrobaeza) [reply partner_second_lastname/models/res_partner.py]: Yes, this way you update multiple fields without triggering a write.

File: partner_second_lastname/models/res_partner.py (hunk 12 above)
Comment (pedrobaeza) [reply partner_second_lastname/models/res_partner.py:63]: The same

File: partner_second_lastname/models/res_partner.py (hunk 11 above)
Comment (lasley) [reply partner_second_lastname/models/res_partner.py]: Interesting, but when is the write actually triggered?

Comment (pedrobaeza) [reply partner_second_lastname/models/res_partner.py]: At the end of the inverse computation round

Comment (pedrobaeza) [reply partner_second_lastname/models/res_partner.py]: This is only when saving. If not, everything is kept in cache.

**Comment by gustavovalverde** [comment]
Is it safe to use this "as is" before being merged?

**Comment by pedrobaeza** [comment]
@gustavovalverde Why don't you try and tell me?

**Review by lasley — APPROVED** [review]

**Comment by lasley** [comment]
@pedrobaeza - squash at your leisure & we can merge

**Comment by pedrobaeza** [comment]
OK, merging then

**Pull request merged**